_posts/2020-04-13-Julia-Antipatterns.md

```diff
@@ -0,0 +1,831 @@
+
+```julia
+str_dict = Dict(string(k)=> v for (k,v) in dict)  # convert all the keys to strings
+@btime $(str_dict)["d"];
+```
+</div>
+
+**Output:**
+
+<div class="jupyter-stream jupyter-cell">
+
+```
+
+  18.077 ns (0 allocations: 0 bytes)
+
+```
+</div>
+
+One alternative is [OrderedCollection's](https://github.com/JuliaCollections/OrderedCollections.jl/) `LittleDict`.
+This is a naive pair of lists based dictionary, with expected time `O(n)` but with a much lower cost per actual element as it doesn't need to `hash` anything.
+
+**Input:**
+
+<div class="jupyter-input jupyter-cell">
+
+```julia
+using OrderedCollections
+little_dict = LittleDict(dict)
+@btime $(little_dict)[:d];
+```
+</div>
+
+**Output:**
+
+<div class="jupyter-stream jupyter-cell">
+
+```
+
+  4.793 ns (0 allocations: 0 bytes)
+
+```
+</div>
+
+It also comes in a immutable form via `freeze` (or by constructing it with `Tuple`s)
+
+**Input:**
+
+<div class="jupyter-input jupyter-cell">
+
+```julia
+frozen_little_dict = freeze(LittleDict(dict))
+@btime $(frozen_little_dict)[:d];
+```
+</div>
+
+**Output:**
+
+<div class="jupyter-stream jupyter-cell">
+
+```
+
+  4.547 ns (0 allocations: 0 bytes)
+
+```
+</div>
+
+But the real winner here is the `NamedTuple`.
+Which performs [constant-folding](https://en.wikipedia.org/wiki/Constant_folding) to remove the indexing operation entirely, and just put the result directly into the AST during compilation.
+
+**Input:**
+
+<div class="jupyter-input jupyter-cell">
+
+```julia
+named_tuple = (; dict...)  # create a NamedTuple with the same content as the dict
+@btime $(named_tuple)[:d];
+```
+</div>
+
+**Output:**
+
+<div class="jupyter-stream jupyter-cell">
+
+```
+
+  0.032 ns (0 allocations: 0 bytes)
+
+```
+</div>
+
+If one is thinking _"I have some constants and I want to group them"_ then look no further than the `NamedTuple`.
+It is the right answer.
+`Dict` is best when you don't know all the keys when you write the code, and/or if values need to be updated.
+
+## Over constraining argument types
+
+Type constraints in Julia are **only** for dispatch.
+If you don't have multiple methods for a function, you don't need any type-constraints.
+If you must add type-constraints (for dispatch) do so as loosely as possible.
+
+### Reasons people do this:
+I think this comes from three main places.
+The belief it would make the code faster (false), safer (mostly false), or easier to understand (true)
+The first two come from different languages which do not act like Julia.
+The last point on ease of understanding is absolutely true, but not worth it most of the time.
+One other reason I can imagine is misunderstanding [this part of the documentation](https://docs.julialang.org/en/v1/manual/performance-tips/), which applies to `struct` fields, not arguments.
+I hope that misunderstanding is not a common reason.
+
+The belief that adding type constraints makes code faster, comes from not understanding how the JIT compiler works.
```

Comment:
First sentence can be phrased in a positive way, indicating that the paragraph is a brief introduction to how the JIT works